sycl/CMakeLists.txt

```diff
@@ -52,31 +52,38 @@ if( NOT OpenCL_INCLUDE_DIRS )
     COMMENT           "Downloading OpenCL headers."
   )
 else()
-  message("OpenCL header have been found under ${OpenCL_INCLUDE_DIRS}.")
   add_custom_target( ocl-headers ALL
     DEPENDS ${OpenCL_INCLUDE_DIRS}
     COMMAND ${CMAKE_COMMAND} -E copy_directory ${OpenCL_INCLUDE_DIRS}/CL ${dst_dir}/CL
     COMMENT "Copying OpenCL headers ..."
   )
 endif()
 
-if( NOT OpenCL_LIBRARIES )
-  message("OpenCL_LIBRARIES is missed. Try to build from GitHub sources.")
-  set(OpenCL_LIBRARIES "${LLVM_LIBRARY_OUTPUT_INTDIR}/libOpenCL.so")
+if( NOT OpenCL_LIBRARIES OR ${OpenCL_VERSION_STRING} VERSION_LESS 2.2 )
+  message("OpenCL_LIBRARY is missed or disregarded. Will try to download OpenCL ICD Loader 2.2 or newer from github.com")
+  if(MSVC)
```

Comment:
Couple notes.
There are dedicated CMake variables for library prefixes/suffixes:
https://gitlab.kitware.com/cmake/community/wikis/doc/cmake/Useful-Variables
CMAKE_ARGS accepts generator expressions. I believe it would be better way to set OPENCL_ICD_LOADER_REQUIRE_WDK that way.

Reply:
I second @valerydmit here.
It looks like we are trying to link with OpenCL library differently depending on the platform.
IIRC, dynamic linking on Windows is slightly different from Linux, but it also should be handled by the cmake just fine. We just need communicate to cmake that "OpenCL is a library" and it shouldn't require low level manipulations with file extensions and compiler flags to link the library.

Reply:
Here are some reasons why find_package(OpenCL) is needed when 
1) If I remove '(OpenCL_INCLUDE_DIR AND OpenCL_LIBRARY) OR ' then OpenCL is NOT FOUND 
   even if I manually set -DOpenCL_INCLUDE_DIR=$OPENCL_HEADERS -DOpenCL_LIBRARY=$OPENCL_LOADER
   
     OpenCL_INCLUDE_DIR is missed. Will try to download OpenCL headers from github.com
     OpenCL_LIBRARY is missed. Will try to download OpenCL ICD Loader from github.com
	 
   That happens because we set OpenCL_INCLUDE_DIR and then check OpenCL_INCLUDE_DIRS below (DIR vs DIRS).
2) find_package(OpenCL) initializes other useful vars like OPENCL_FOUND, etc.

Regarding your concern "their values might be reset(!) by this function".
I tried that. It does NOT re-define what you explicitly defined. OpenCL_INCLUDE_DIR has the highest priority:
a) "-DOpenCL_INCLUDE_DIR=garbage -DOpenCL_LIBRARY=garbage" ==> OpenCL is FOUND, not even checked.
-- Looking for CL_VERSION_2_2
-- Looking for CL_VERSION_2_2 - not found
-- Looking for CL_VERSION_2_1
-- Looking for CL_VERSION_2_1 - not found
-- Looking for CL_VERSION_2_0
-- Looking for CL_VERSION_2_0 - not found
-- Looking for CL_VERSION_1_2
-- Looking for CL_VERSION_1_2 - not found
-- Looking for CL_VERSION_1_1
-- Looking for CL_VERSION_1_1 - not found
-- Looking for CL_VERSION_1_0
-- Looking for CL_VERSION_1_0 - not found
-- Found OpenCL: D:/iusers/<username>/ws/gs_public/build_ninja/garbage   
   
b) "-DBUILD_WITH_SYSTEM_OPENCL=ON -DOpenCL_INCLUDE_DIR=garbage -DOpenCL_LIBRARY=garbage" ==> SAME OUTPUT AS IN (a) above).

Reply:
1. Should we just unify naming across the whole project?
2. Do you have OpenCL installed on you system? I.e. what will happen if these variables are not set?

Reply:
> 1. Should we just unify naming across the whole project?

I used the name 'BUILD_WITH_SYSTEM_OPENCL' suggested by you. Should I change it to OpenCL_BUILD_WITH_SYSTEM_SDK ?

Reply:
> 2\. Do you have OpenCL installed on you system? I.e. what will happen if these variables are not set?

If OpenCL_INCLUDE_DIR and OpenCL_LIBRARY are not set, then 

a) if BUILD_WITH_SYSTEM_OPENCL unset, then nothing is found, and they are downloaded/built from github. 

b) if BUILD_WITH_SYSTEM_OPENCL is set then SYSTEM SDK may be found. I have OpenCL headers 2.1 on my lab machine, it is found and tried to be used. 
BTW, currently SYCL compiler cannot be built with 2.1 headers because we include both files cl_ext.h and cl_ext_intel.h and they define same structures in 2.1, but not in 2.2.

Reply:
> > 1. Should we just unify naming across the whole project?

> I used the name 'BUILD_WITH_SYSTEM_OPENCL' suggested by you. Should I change it to OpenCL_BUILD_WITH_SYSTEM_SDK ?

No.

Sorry I wasn't clear enough. I was talking about OpenCL_INCLUDE_DIR/OpenCL_INCLUDE_DIRS and OpenCL_LIBRARY/DOpenCL_LIBRARIES.

Anyway. Let's fix more important issues first and refactor it later.